junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ConcurrentHierarchicalTestExecutorService.java

```diff
@@ -0,0 +1,729 @@
+	}
+
+	public ConcurrentHierarchicalTestExecutorService(ParallelExecutionConfiguration configuration) {
+		this(configuration, ClassLoaderUtils.getDefaultClassLoader());
+	}
+
+	ConcurrentHierarchicalTestExecutorService(ParallelExecutionConfiguration configuration, ClassLoader classLoader) {
+		ThreadFactory threadFactory = new WorkerThreadFactory(classLoader);
+		threadPool = new ThreadPoolExecutor(configuration.getCorePoolSize(), configuration.getMaxPoolSize(),
+			configuration.getKeepAliveSeconds(), SECONDS, new SynchronousQueue<>(), threadFactory);
+		parallelism = configuration.getParallelism();
+		workerLeaseManager = new WorkerLeaseManager(parallelism, this::maybeStartWorker);
+		LOGGER.trace(() -> "initialized thread pool for parallelism of " + configuration.getParallelism());
+	}
+
+	@Override
+	public void close() {
+		LOGGER.trace(() -> "shutting down thread pool");
+		threadPool.shutdownNow();
+	}
+
+	@Override
+	public Future<@Nullable Void> submit(TestTask testTask) {
+		LOGGER.trace(() -> "submit: " + testTask);
+
+		var workerThread = WorkerThread.get();
+		if (workerThread == null) {
+			return enqueue(testTask).future();
+		}
+
+		if (testTask.getExecutionMode() == SAME_THREAD) {
+			workerThread.executeTask(testTask);
+			return completedFuture(null);
+		}
+
+		return new WorkStealingFuture(enqueue(testTask));
+	}
+
+	@Override
+	public void invokeAll(List<? extends TestTask> testTasks) {
+		LOGGER.trace(() -> "invokeAll: " + testTasks);
+
+		var workerThread = WorkerThread.get();
+		Preconditions.condition(workerThread != null && workerThread.executor() == this,
+			"invokeAll() must be called from a worker thread that belongs to this executor");
+
+		workerThread.invokeAll(testTasks);
+	}
+
+	private WorkQueue.Entry enqueue(TestTask testTask) {
+		var entry = workQueue.add(testTask);
+		maybeStartWorker();
+		return entry;
+	}
+
+	private void forkAll(Collection<WorkQueue.Entry> entries) {
+		if (entries.isEmpty()) {
+			return;
+		}
+		workQueue.addAll(entries);
+		// start at most (parallelism - 1) new workers as this method is called from a worker thread holding a lease
+		for (int i = 0; i < Math.min(parallelism - 1, entries.size()); i++) {
+			maybeStartWorker();
+		}
+	}
+
+	private void maybeStartWorker() {
+		maybeStartWorker(() -> false);
+	}
+
+	private void maybeStartWorker(BooleanSupplier doneCondition) {
+		if (threadPool.isShutdown() || workQueue.isEmpty() || doneCondition.getAsBoolean()) {
+			return;
+		}
+		var workerLease = workerLeaseManager.tryAcquire();
+		if (workerLease == null) {
+			return;
+		}
+		try {
+			threadPool.execute(() -> {
+				LOGGER.trace(() -> "starting worker");
+				try {
+					WorkerThread.getOrThrow().processQueueEntries(workerLease, doneCondition);
+				}
+				finally {
+					workerLease.release(false);
+					LOGGER.trace(() -> "stopping worker");
+				}
+				maybeStartWorker(doneCondition);
```

Comment:
Note to self: I still can't quite get my head around this line.

If there is a race condition where `processQueueEntries` may run out of tasks, and `forkAll` adding tasks and starting new workers, but failing because this worker is still active.

Then I would expect `processQueueEntries` to read from the work queue in a blocking manner.

Reply:
Answer: Because we're managing the degree of parallelism ourselves, blocking here will occupy a lease.